### ISSUE TITLE
Convergence lab: fixed-scale + warmstart validation notebook

### ISSUE THREAD

**Issue by jaredgalloway** [matsengrp/multidms#246]
# Convergence lab: fixed-scale + warmstart validation notebook

**Status:** ready · **Type:** task (experiment + small code change) · **Repo:** multidms

## Goal

Build a fast, local artifact that reproduces the scv2-spike convergence
pathology on full real data and prescribes the fix, scored on the metric we
actually care about — replicate-shift correlation — not just iteration counts.
Land the convergence fix (`recompute_scale` toggle) in `jaxmodels` as a real,
reviewable parameter so the experiment exercises mainline code, then validate it
end-to-end in a marimo notebook with a seconds-to-minutes feedback loop.

This replaces the hours-long remote pipeline as the inner loop for diagnosing
modeling/optimizer problems.

## Background: what we confirmed by poking at the data and code

The user reported four concerning results from the dashboard
(`results-prod-235-times-seen-threshold`). Cross-checking each against the data
and `jaxmodels.py`:

| Reported observation | Finding | Verdict |
|---|---|---|
| Loss barely decreases / sometimes increases | `loss_per_variant` flat at ~0.22 across all 50 iters | **Confirmed** |
| Fusionreg drives where β0/α converge | β0 climbs and α plummets to ≈ −5, fanned by fusionreg, still moving at iter 50 | **Confirmed** |
| objective_error never reaches ~1e-6 | mostly 1e-3–1e-4, noisy | **Confirmed** |
| Delta's GE-landscape floor sits below BA1/BA2 | This is a **model-predicted** floor, not a data property. The measured `func_score` is clamped to −3.5/+2.5 in **data prep** (a prep step on the input targets, not a model bound), so the input data has no per-condition floor difference. Delta is the *least* depleted condition (median −0.64 vs −1.2/−1.3) and has ~half the data (55.8k vs ~118–129k variants). | **Reframed: it's a predicted-floor divergence, not a data floor** |

The Delta "floor below the others" in the GE-landscape plot is a **model
artifact** of the fit, not a property of the input data. `α` scales the sigmoid
*output*: `predict_score = α·(σ(φ(X)) − σ(φ(x_wt)))` (jaxmodels.py:255). With
`α ≈ −5` and Delta's β0 drifting, Delta's **predicted** fitness floor sinks to
≈ −5, well below the *predicted* floors of BA1/BA2. (The −3.5/+2.5 data-prep
clamp is on the input targets and does not bound predictions, so it is
irrelevant to this predicted-floor comparison.) This is the `(α, β0)` sigmoid
degeneracy that the #213 → #215 → #218 chain has been chasing.

### The root cause is in mainline code, verbatim

A **post-hoc** ablation (run after `results-prod-235` as a #213 follow-up;
`experiments/scv2-spike/results/ablation-scale-vs-ridge/FINDINGS.md`) concluded:
revert to a **fixed `scale` computed once** per `fit()`. The current mainline
`fit()` still recomputes `scale` every outer sweep (jaxmodels.py:631–644):

```python
raw_obj = float(abs(objective_total(..., scale=1.0, beta0_ridge=beta0_ridge)))
scale = raw_obj if raw_obj > 1e-30 else 1.0
obj_old = raw_obj / scale  # 1.0 by construction
```

Because `scale ≈ raw_obj`, `obj_old ≡ 1.0` every iteration, so
`objective_error = |obj_old − obj| / max(...)` can never reliably hit `block_tol
= 1e-6`. And the lasso threshold `fusionreg / scale` drifts each sweep → a
non-stationary proximal problem → the β0/α fan-out by fusionreg. The ablation
proved fixed-scale converges in all four ridge-0 cells while recompute-scale
failed three of four — but it measured only iteration counts, never replicate
reproducibility, and it used a throwaway worktree toggle that was reverted.

## Deliverables

This spec produces **two coupled deliverables**.

### Deliverable A — `recompute_scale` toggle in jaxmodels (the permanent fix)

- Add `recompute_scale: bool = True` to `jaxmodels.fit()`. Default `True`
  **preserves current behavior** → non-breaking, CI-safe.
- When `False`: compute `scale` **once before the outer loop** (after warmstart
  if `warmstart=True`, so the constant is computed near a solution), hold it
  constant for the lifetime of the fit, and set `obj_old` from the *previous
  sweep's* scaled objective so `objective_error` measures a true relative change
  (a well-posed Cauchy criterion) rather than `|1.0 − obj|`.
- **Plumb the flag through four explicit signatures** — there is no clean
  kwargs passthrough (`fit_one_model`'s `**kwargs` are stored as bookkeeping,
  not forwarded to `model.fit`). The four edit sites:
  1. `jaxmodels.fit` signature (`jaxmodels.py:364`) + the scale/`obj_old` logic
     (`jaxmodels.py:631–644`) and the `objective_error` line (`jaxmodels.py:778`).
  2. `Model.fit` signature (`model.py:151`) + its forwarding call (`model.py:240`).
  3. `fit_one_model` signature (`model_collection.py:61`) + its `model.fit(...)`
     call (`model_collection.py:159`).
  4. `fit_models_path` needs no change: `_fit_one_path_step`
     (`model_collection.py:447`) already forwards `**kwargs` to `fit_one_model`,
     so it passes through once (3) accepts the param.
- **Tests** (mirror the prior `ablation.py` reference implementation, which
  already encoded these invariants):
  - **T1 (scale constant):** fit a small 2-condition dataset with
    `recompute_scale=False, block_iters≥5`; assert `scale` is bit-for-bit
    identical across all sweeps. With `recompute_scale=True`, assert it varies
    (so the test discriminates).
  - **T2 (true relative change):** with `recompute_scale=False`, assert
    `objective_error` equals `|obj_prev − obj| / max(|obj_prev|, |obj|, 1)` on
    the scaled objective with `obj_old` taken from the *previous sweep* — i.e.
    NOT the `|1.0 − obj|` the recompute path produces by construction
    (`jaxmodels.py:644` + `:778`).
  - **T3 (regression guard):** fit with `recompute_scale=True` (default) and
    assert the trajectory matches the pre-change output bit-for-bit (omitting
    the kwarg ≡ passing `True`). Guards the non-breaking claim.
  - Existing tests must pass unchanged. `summary.txt`'s
    `scale=fixed … fr=8.0e-05 iter=19 conv=True err=3.9e-07 inc=5` is a
    ready-made expected-value anchor for an integration smoke test.
- Do **not** change the default in this work, and do **not** touch the
  early-stopping gate or `beta0_ridge` — those are separate concerns. The
  notebook's job is to produce the evidence that justifies flipping the default
  in a follow-up.

### Deliverable B — `convergence_lab.py` marimo notebook (the validation)

Location: `experiments/scv2-spike/notebooks/convergence_lab.py`. This is the
first marimo `.py` app under that dir (the siblings are `.ipynb` + `_common.py`);
follow the style of the existing marimo dashboard at `experiments/dashboard.py`,
not the sibling notebooks. Use the `bip-marimo` skill when authoring.

Data substrate: **full** real
`results-prod-235-times-seen-threshold/training_functional_scores.csv`, both
replicates, all three conditions. No subsampling (eliminates subsampling-artifact
risk). Fits are cached to pickle so re-runs are instant.

## The factorial

Per replicate (run rep_1 and rep_2 so shifts can be correlated):

```
   scale     ∈ {recompute, fixed}        ← the convergence fix under test
 ⊗ warmstart ∈ {on, off}                  ← does Ridge warmstart speed convergence
                                             WITHOUT hurting reproducibility?
 ⊗ fusionreg ∈ {0.0, 8e-5, 3.2e-4}        ← spans prod grid; 8e-5 = ablation's worst
                                             regression cell, 3.2e-4 = hardest α fan-out
 = 2 × 2 × 3 = 12 cells per replicate × 2 replicates = 24 fits
```

**Held fixed:** `share_alpha=True` (prod default), Sigmoid GE, all three
conditions, `beta0_ridge=0.0` (prod effectively abandoned it),
`block_tol=1e-6`, `block_iters` capped at 50 (the value the prior ablation used;
enough to see fixed-scale converge before the cap and recompute not).

**Layered diagnostic probe (NOT a full axis):** an optional `α`-bound/clamp
applied only to the best-converging cell, to show whether clamping α away from
the −5 collapse lifts Delta's predicted floor back into line with the other
conditions' predicted floors. One switch in the diagnostic panel, not 24× more
fits.

Three fusionreg values (not the full 9) keep the sweep to ~20–30 min on full
~300k-variant data while bracketing the pathology; caching makes re-runs instant.

### ① Convergence (the ablation's axis)

Per cell: `converged` (objective_error < block_tol before cap), final
`objective_error`, iters-to-converge, and `inc` = number of outer iterations
where the objective *increased* (oscillation proxy). These are **derived** from
the fit trajectory df (recorded by #212/#221/#228; exposed via
`Model.convergence_trajectory_df`), not stored columns. **Reuse the exact
derivation in `ablation.py:72–81`** so `inc` matches the existing summary's
`n_obj_increase`:
`converged = traj["objective_error_trajectory"].iloc[-1] < block_tol`;
`iters = len(traj)`;
`inc = (np.diff(objective_total_trajectory) > 1e-9 * |obj[:-1]|).sum()`.

### ② Replicate-shift correlation (the ultimate validation) ⭐

**Reuse the existing canonical convention** — do not reinvent it. Build a
`ModelCollection` per `(scale, warmstart)` cell and call
`ModelCollection.mut_param_dataset_correlation(x="fusionreg", return_data=True)`
(model_collection.py:1346). It already pairs datasets (rep_1, rep_2) via
`itertools.combinations`, correlates per `mut_param` across mutations, and emits
the exact schema of the existing `library_replicate_correlation.csv`:

```
datasets,mut_param,correlation,fusionreg
"rep_1,rep_2",shift_Delta,0.62...,0.0
```

**Filter to the shift params for the headline metric.** The method returns rows
for *all* mut_params (`beta_Delta`, `beta_Omicron_BA1`, `beta_Omicron_BA2`,
`shift_Delta`, `shift_Omicron_BA2`, `predicted_func_score_*`). Metric ② is
**replicate-shift** correlation, so Cell 3 filters to
`mut_param.str.startswith("shift")` (only non-reference conditions Delta and BA2
have shifts; BA1 is the reference). Use `r=1` (Pearson) to match the existing
CSV. Tag each row with its `(scale, warmstart)` cell.

## Notebook wiring

```
convergence_lab.py  (marimo, reactive)
├─ Cell 1  load full results-prod-235 training data → multidms.Data per (replicate, condition group)
├─ Cell 2  FACTORIAL runner over scale × warmstart × fusionreg × replicate
│            via fit_models() with recompute_scale passthrough; cache to pickle
├─ Cell 3  ModelCollection per (scale,warmstart) → mut_param_dataset_correlation(return_data=True)  ⇒ Metric ②
├─ Cell 4  Metric ① table: converged / final obj_error / iters / inc, per cell
├─ Cell 5  DECISIVE plot: correlation vs fusionreg, one line per (scale × warmstart)
├─ Cell 6  Diagnostic panel (reactive selector): α-trajectory · β0-drift · per-condition predicted-floor (Delta vs BA1/BA2)
└─ Cell 7  α-bound probe (optional switch) on the best cell → does clamping α remove the Delta floor artifact?
```

**Reuse, not reinvention:**
- ② correlation → `mut_param_dataset_correlation` (exists)
- fitting → `model_collection.fit_models` (exists; gains `recompute_scale` passthrough)
- α/β0/obj_error trajectories → already recorded by convergence diagnostics (#212/#221/#228)
- GE landscape / Delta floor → reuse the dashboard's GE-landscape plot logic

## Success criteria

1. **Reproduces the pathology:** recompute-scale → flat loss, α → −5, obj_error
   stuck at 1e-3–1e-4, on full real data.
2. **Fix works mechanically:** fixed-scale → loss decreases, obj_error reaches
   ~1e-6, α stays sane.
3. ⭐ **Fix works scientifically:** replicate-shift correlation (canonical
   metric) is **≥ as high** under fixed-scale as recompute at every fusionreg
   and **does not collapse** at high fusionreg.
4. **Warmstart is additive, not costly:** warmstart reduces iters-to-converge
   **without lowering** replicate correlation.
5. **Degeneracy is localized:** diagnostic shows Delta's predicted floor tracks
   α; the α-bound probe lifts Delta's predicted floor back into line with the
   BA1/BA2 predicted floors.

> **The fork this resolves:** if criterion 3 *fails* — fixed-scale converges but
> replicate correlation stays bad — that is the signal the problem is not just a
> scale bug but a deeper framework issue (the open #243 concurrent-fitting
> frontier). Either way, a fast local experiment localizes it instead of hours
> on the remote pipeline.

## Out of scope

- Flipping the `recompute_scale` default (a follow-up, justified by this
  notebook's evidence).
- Changing the early-stopping gate or `beta0_ridge` (separate concerns from
  #215).
- The concurrent/single-solve fitter (#243) — this experiment informs whether
  that frontier is needed, but does not implement it.
- Synthetic ground-truth data — we rely on replicate correlation as the
  reproducibility metric, per the user's "ultimate validation."

## References

- Ablation findings: `experiments/scv2-spike/results/ablation-scale-vs-ridge/FINDINGS.md`
- Ablation harness (reference impl of Deliverable A + metric ① derivations,
  `ablation.py:72–81`): `experiments/scv2-spike/results/ablation-scale-vs-ridge/ablation.py`
- Convergence trajectory accessor: `Model.convergence_trajectory_df` (`model.py:130`)
- Canonical replicate-correlation CSV schema (`datasets, mut_param, correlation, fusionreg`):
  `experiments/scv2-spike/results-prod-235-times-seen-threshold/library_replicate_correlation.csv`
- Mainline scale recompute: `multidms/jaxmodels.py:631–644`
- α in sigmoid: `multidms/jaxmodels.py:255`
- Canonical replicate correlation: `multidms/model_collection.py:1346`
- Prior PRs: #215 (dynamic scale / stopping / beta0_ridge), #218 (share alpha)
- Related open issues: #243/#244/#245 (concurrent fitting), #241 (reduced block
  iters), #242 (local-minima trapping), #240 (path fitter silent failures)
